File: westcoast/src/main/java/frc/commands/targeting/SetupShot.java
package frc.commands.targeting;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.robot.definition.Definition;
import frc.subsystems.DriveTrainSubsystem;
import org.jetbrains.annotations.NotNull;

public final class SetupShot extends CommandBase {
    @NotNull
    private final PIDController turn;
    @NotNull
    private final PIDController forward;
    @NotNull
    private final DriveTrainSubsystem drive;
    private final double distance;
    private int turnOnTarget;
    private int forwardOnTarget;
    private int targetLost;
    private boolean hasInitialTarget;

    public SetupShot(
            @NotNull Definition definition,
            double distance
    ) {
        this.turn = definition.controllers.alignToTarget();
        this.forward = definition.controllers.forwardAlignToTarget();
        this.drive = definition.subsystems.driveTrain;
        this.distance = distance;
    }

    @Override
    public void initialize() {
        turn.reset();
        forward.reset();
        turnOnTarget = 0;
        forwardOnTarget = 0;
        targetLost = 0;
        hasInitialTarget = true;//drive.vision.hasTarget();
    }

    @Override
    public void execute() {
        double forwardValue = forward.calculate(drive.vision.getDistance(), distance);
        double turnValue = -turn.calculate(drive.vision.getXOffset(), 0.0);
        drive.arcade(
                forwardValue,
                turnValue,
                false
        );
        turnOnTarget = turn.atSetpoint() ? (turnOnTarget + 1) : 0;
        forwardOnTarget = forward.atSetpoint() ? (forwardOnTarget + 1) : 0;
        targetLost = !drive.vision.hasTarget() ? (targetLost + 1) : 0;
    }

    @Override
    public void end(boolean interrupted) {
        drive.stop();
    }

    @Override
    public boolean isFinished() {
        return ((turnOnTarget > 10) && (forwardOnTarget > 10)) || (!hasInitialTarget) || (targetLost > 5);
    }
}

File: westcoast/src/main/kotlin/frc/commands/targeting/SetupShot.kt
package frc.commands.targeting

import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.wpilibj2.command.CommandBase
import frc.subsystems.DriveTrainSubsystem

class SetupShot(
    private val turn: PIDController,
    private val forward: PIDController,
    private val drive: DriveTrainSubsystem,
    private val distance: () -> Double
) : CommandBase() {

    private var turnOnTarget = 0
    private var forwardOnTarget = 0
    private var targetLost = 0
    private val hasTarget = drive.vision.hasTarget

    override fun initialize() {
        turn.reset()
        forward.reset()
        turnOnTarget = 0
        forwardOnTarget = 0
        targetLost = 0
    }

    override fun execute() {
        drive.arcade(
            forward.calculate(drive.vision.distance, distance()).coerceIn(-0.1, 0.1),
            -turn.calculate(drive.vision.xOffset, 0.0).coerceIn(-0.2, 0.2),
            false
        )
        turnOnTarget = if (turn.atSetpoint()) turnOnTarget + 1 else 0
        forwardOnTarget = if (forward.atSetpoint()) forwardOnTarget + 1 else 0
        targetLost = if (drive.vision.hasTarget) targetLost + 1 else 0
    }

    override fun end(interrupted: Boolean) {
        drive.stop()
    }

    override fun isFinished(): Boolean {
        return turnOnTarget > 10 && forwardOnTarget > 10 || !hasTarget || targetLost > 5
    }
}